### PR TITLE
added support to bme280 for humidity. code based on bosch driver

### PR DESCRIPTION
**Pull request by Mircho**
Hi Pim and Sergey,
I have been using the homeassistant lib successfully on my sensor nodes. The thing I was missing was the humidity support for BME280 Bosch sensors and I added it by combing through the datasheet and the driver code coming from Bosch.
Hope this helps

### CODE REVIEW

File: src/mgos_barometer_bme280.c [hunk 1]
```diff
@@ -74,19 +79,51 @@ bool mgos_barometer_bme280_create(struct mgos_barometer *dev) {
...
+    bme280_data->calib.dig_H4 = dig_h4_msb | dig_h4_lsb;
+    dig_h5_msb = (int16_t)(int8_t)reg_data[5] * 16;
+    dig_h5_lsb = (int16_t)(reg_data[4] >> 4);
+    bme280_data->calib.dig_H5 = dig_h5_msb | dig_h5_lsb;
+    bme280_data->calib.dig_H6 = (int8_t)reg_data[6];
+  }
+
   // SPI | 0.5ms period | 16X IIR filter
-  if (!mgos_i2c_write_reg_b(dev->i2c, dev->i2caddr, BME280_REG_CONFIG, 0x00 | BME280_STANDBY_500us << 2 | BME280_FILTER_16X << 5)) {
+  if (!mgos_i2c_write_reg_b(dev->i2c, dev->i2caddr, BME280_REG_CONFIG, 0x00 | BME280_STANDBY_500us << 2 | BME280_FILTER_2X << 5)) {
```
Comment (pimvanpelt) [inline src/mgos_barometer_bme280.c]: why change to 2x filter here?

File: src/mgos_barometer_bme280.c [hunk 2]
```diff
@@ -155,6 +194,22 @@ bool mgos_barometer_bme280_read(struct mgos_barometer *dev) {
   }
   dev->pressure = (float)P;
 
+// Compensatoin for humidity
```
Comment (pimvanpelt) [inline src/mgos_barometer_bme280.c]: spelling

File: src/mgos_barometer_bme280.c [hunk 3]
```diff
@@ -17,6 +17,8 @@
 #include "mgos_barometer_bme280.h"
 #include "mgos_i2c.h"
 
+#define BME280_CONCAT_BYTES(msb, lsb)            (((uint16_t)msb << 8) | (uint16_t)lsb)
+
 // Datasheet:
 // https://cdn-shop.adafruit.com/datasheets/BST-BME280_DS001-10.pdf
 //
```
Comment (pimvanpelt) [inline src/mgos_barometer_bme280.c:24]: You can remove the TODO here!

File: src/mgos_barometer_bme280.c [hunk 4]
```diff
@@ -33,6 +35,9 @@ bool mgos_barometer_bme280_detect(struct mgos_barometer *dev) {
     return false;
   }
 
+  LOG(LL_INFO, ("Detecting BMx280 %d, reg b is: %d", dev->i2caddr, val));
```
Comment (pimvanpelt) [inline src/mgos_barometer_bme280.c]: I'm not convinced this is a useful logline for users. Perhaps make it LL_DEBUG ?

File: src/mgos_barometer_bme280.c [hunk 5]
```diff
@@ -116,17 +153,19 @@ bool mgos_barometer_bme280_read(struct mgos_barometer *dev) {
   }
 
   // read data from sensor
-  uint8_t data[6];
-  if (!mgos_i2c_read_reg_n(dev->i2c, dev->i2caddr, BME280_REG_PRESSURE_MSB, 6, data)) {
+  uint8_t data[8];
```
Comment (pimvanpelt) [inline src/mgos_barometer_bme280.c:156]: I suppose the answer will be 'yes', but have you asserted that it's safe to read 8 bytes from a BMP280 which doesn't have these registers for `Hadc` ?

File: src/mgos_barometer_bme280.c [hunk 6]
```diff
@@ -155,6 +194,22 @@ bool mgos_barometer_bme280_read(struct mgos_barometer *dev) {
   }
   dev->pressure = (float)P;
 
+// Compensatoin for humidity
+  var1 = ((double)t_fine) - 76800.0;
```
Comment (pimvanpelt) [inline src/mgos_barometer_bme280.c]: Can you execute this block only if the capability is set.

File: src/mgos_barometer_bme280.c [hunk 7]
```diff
@@ -155,6 +194,22 @@ bool mgos_barometer_bme280_read(struct mgos_barometer *dev) {
...
+  var6 = var3 * var4 * (var5 * var6);
+  H = var6 * (1.0 - ((double) bme280_data->calib.dig_H1) * var6 / 524288.0);
+
+  if (H > H_max)
+      H = H_max;
+  else if (H < H_min)
+      H = H_min;
+
+  dev->humidity = (float)H;
 //  LOG(LL_DEBUG, ("P=%.2f T=%.2f", dev->pressure, dev->temperature));
```
Comment (pimvanpelt) [inline src/mgos_barometer_bme280.c]: I know the line is commented out, but can you add humidity to the LL_DEBUG here ?

File: src/mgos_barometer_bme280.h [hunk 8]
```diff
@@ -22,6 +22,7 @@
 // DevID: 0x56/0x57 are samples of BMP280; 0x58 is mass production BMP280; 0x60 is BME280
 #define BME280_REG_DEVID                           (0xD0) /* Chip ID Register */
 #define BME280_REG_RESET                           (0xE0) /* Softreset Register */
+#define BME280_REG_CTRL_HUM                       (0xF2) /* Humidity oversampling */
```
Comment (pimvanpelt) [inline src/mgos_barometer_bme280.h]: nit: alignment of the value, missing one space :)

**Review by pimvanpelt — CHANGES_REQUESTED** [review]
Great work! The only thing I'd ask is that you assert that this driver now still works with BMP280 which misses the hygrometer. Specifically curious to ensure that the 8 byte read is safe - possibly you could make the read be 6 bytes if the capability isn't set, and 8 bytes if it is. 

Otherwise, all I found were a few nits and cosmetic issues. Thanks for the contribution!

File: src/mgos_barometer_bme280.c (hunk 1 above)
Comment (Mircho) [reply src/mgos_barometer_bme280.c]: Don't recollect why so I would not insist on it.

File: src/mgos_barometer_bme280.c (hunk 3 above)
Comment (Mircho) [reply src/mgos_barometer_bme280.c:24]: Done

File: src/mgos_barometer_bme280.c (hunk 5 above)
Comment (Mircho) [reply src/mgos_barometer_bme280.c:156]: In the "official" Bosch driver they do read 8 bytes at once. I would test this on a BMP280 once I can build a dev board with it.

File: src/mgos_barometer_bme280.c (hunk 6 above)
Comment (Mircho) [reply src/mgos_barometer_bme280.c]: Done

**Comment by Mircho** [comment]
I've amended according to the comments and latest master of my fork has them. You are right that I should test it with BMP280 to fully validate. It might take few days.

File: src/mgos_barometer_bme280.c (hunk 5 above)
Comment (Mircho) [reply src/mgos_barometer_bme280.c:156]: Tested with BMP280:

```
[Aug 27 00:03:00.057] mgos_rpc_channel_mq:203 0x3ffc9200 GarageSensor_B69170/rpc
[Aug 27 00:03:00.060] mgos_rpc_channel_ua:313 0x3ffc9670 UART0
[Aug 27 00:03:00.123] main.c:12               Registering Home assistant from file: ha.json
```

`GarageSensor_B69170/sensor/environment {"pressure":948.25,"temperature":25.63}
`
```json
{
    "provider" : {
       "gpio" : [
          { "name" : "builtinled", "type": "switch", "invert": true, "gpio": 22 },
       ],
        "barometer" : [
            { "name" : "environment", "i2caddr" : 118, "period" : 60, "type" : "bmp280" }
        ]
     }
}
```

It took me some time to dig out a working BMP280 from the pile.

**Pull request merged**